CSUB-386: Enable cargo audit in CI

atodorov opened the pull request:
Description of proposed changes:


Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

atodorov committed:
Enable cargo audit in CI

CAGS295 commented on .github/workflows/ci.yml:
```diff
@@ -34,6 +34,39 @@ jobs:
+
+      - name: Install cargo audit
+        run: |
+          cargo install cargo-audit --features=fix
```
It doesn't look like we are using the opted in fix feature.

atodorov replied:
That's correct. I wasn't sure if we wanted this to automatically commit to GitHub or not, especially given that it doesn't fix all issues that it finds.

atodorov committed:
cargo audit fix

atodorov committed:
Ignore RUSTSEC-2020-0071

Potential segfault in the time crate:
https://rustsec.org/advisories/RUSTSEC-2020-0071

this is not a direct dependency of creditcoin-node and will be fixed
via a future upgrade of Substrate.

ALSO don't error out on warnings because all of them are coming from
Substrate as well.

atodorov committed:
Make sure to fail for 1st level dependencies of Creditcoin

this is an extra check which should hopefully inform us on warnings
or anything else that is a 1st level dependency of Creditcoin.

AdaJane reviewed (approved):
I'm not crazy about ignoring all indirect dependencies, but I think this is still a nice addition to the pipeline.

Edit: After clarification, everything makes sense

nathanwhit commented on .github/workflows/ci.yml:
```diff
@@ -64,6 +64,50 @@ jobs:
+
+      - name: Run cargo audit
+        run: |
+          cargo audit --ignore RUSTSEC-2020-0071
+          CC_MATCH=$(cargo audit --ignore RUSTSEC-2020-0071 --color never | egrep "^.── creditcoin" || true)
```
Why do we run it twice?

atodorov replied:
First time reports on all possible dependencies, direct and indirect but doesn't fail on warnings like unmaintained crates. 

The second time will explicitly fail if we have anything reported as a 1st level dependency to Creditcoin. For example unknowingly we start using an unmaintained crate.

nathanwhit commented on .github/workflows/ci.yml:
```diff
@@ -64,6 +64,50 @@ jobs:
+          cargo audit --ignore RUSTSEC-2020-0071
+          CC_MATCH=$(cargo audit --ignore RUSTSEC-2020-0071 --color never | egrep "^.── creditcoin" || true)
+
+          # fail for anything which is a 1st level dependency of Creditcoin
+          if [ -n "$CC_MATCH" ]; then
+              echo "FAIL: Direct dependencies of Creditcoin found in audit report"
+              echo "INFO: Inspect the output above for more information"
+              exit 1
+          else
+              echo "INFO: No direct dependencies of Creditcoin found in audit report"
+              exit 0
+          fi
```
I second Ada here, not sure I like ignoring indirect dependencies. I feel like if an advisory is really unactionable we can just add it to the whitelist case by case.

atodorov replied:
> not sure I like ignoring indirect dependencies.

We are not ignoring dependencies, on the contrary.  

The first check will fail regardless where the dependency comes from if there is a vulnerability (excluding vulnerabilities which we explicitly decided to ignore). 

The second check will fail even for warnings if they are reported against 1st level dependencies of Creditcoin.

nathanwhit replied:
Ahh okay I see

nathanwhit reviewed (commented):
A couple questions/comments
